compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -755,7 +755,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             candidate.match_pairs.sort_by_key(|pair| matches!(pair.test_case, TestCase::Or { .. }));
         }
 
-        ret
+        ret.map(|branch| (branch, matched_span))
```

Comment:
Since you add the span regardless of `fully_matched`, some conditions will be duplicated; this may or may not be what you want. For example:
```rust
match x {
    (10.., true) => {}
    (100.., false) => {}
    _ => {}
}
```
In that example, the first condition we try is `10..`, which also matches `100..` but not fully_matches it. So `100..` will show up twice: once when we test if `n>=10`, and the second time when we test if `n>=100`.

Intuitively I'd expect that we only want the second time; what do you think?

Reply:
This is handled in `PatternDecisionCtx::visit_matching_arms` in `mcdc.rs`, I use `false_matched` to indicate if that target is `full_matched` or not and find the right block for it.

Reply:
I should clarify it a bit clearer.
The motivation to catch conditions not `full_matched` is I want to reconstruct decision tree in mcdc.
Say the code is
```rust
// arr: &[i32]
match arr {
    [one] => { /*...*/ }
    [_, two] => { /*...*/ }
    _ => { /*..*/ }
}
```
First we check whether `arr` matches `[one]`. By engaging "partial matched" pattern `[_, two]`, we know condition `[one]` has a false next test `[_, two]`. If we only pass `full_matched` targets to mcdc builder we have no idea how `[one]` is connected to `[_, two]`, true next or false next.

(I forgot to add tests to check this kind of code. Thanks for the reminder now a simple test is added and it help me fix a bug)

Reply:
Oh, I see! If you're reconstructing the whole decision tree, could it be easier to add enough `Span` information to the MIR so that you can act directly on the MIR? I don't know how easy that would be but it feels conceptually simpler. For example we could add info to `SwitchInt` to indicate what pattern it comes from (if any).

Reply:
Indeed that may help a lot. Nevertheless I'm afraid that it results in uncomfortable side effects. Though mcdc is meaningful in some areas, mostly it is required by embedded systems working for safety critical components, which comprise only a very small portion of all applications. I'd tend to encapsulate complexity in mcdc and minimize its effects on other parts that are much more widely depended on.

Besides, mcdc criteria for pattern matching has not reached a consensus within industries yet (due to C is the only practical choice in these areas in the past). The basic idea behind this implementation is addressed by a paper from [ferrous system](https://ferrous-systems.com/ferrocene/) but still there might be other opinions worth consideration. So we'd better keep it in an easy-to-change way.